engine/execution/storehouse/register_store.go

```diff
@@ -0,0 +1,266 @@
+package storehouse
+
+import (
+	"errors"
+	"fmt"
+
+	"go.uber.org/atomic"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+type RegisterStore struct {
+	memStore   *InMemoryRegisterStore
+	diskStore  execution.OnDiskRegisterStore
+	wal        execution.ExecutedFinalizedWAL
+	finalized  execution.FinalizedReader
+	log        zerolog.Logger
+	finalizing *atomic.Bool // making sure only one goroutine is finalizing at a time
+}
+
+var _ execution.RegisterStore = (*RegisterStore)(nil)
+
+func NewRegisterStore(
+	diskStore execution.OnDiskRegisterStore,
+	wal execution.ExecutedFinalizedWAL,
+	finalized execution.FinalizedReader,
+	log zerolog.Logger,
+) (*RegisterStore, error) {
+	// replay the executed and finalized blocks from the write ahead logs
+	// to the OnDiskRegisterStore
+	height, err := syncDiskStore(wal, diskStore, log)
```

Comment:
Right, I thought about this as well. In the happy case, usually, it only has at most 1 block of updates to be synced, which should be very fast. I will leave it as is for now, since the syncDiskStore is doing nothing as we haven't implemented WAL yet. But I will keep this in mind.

Reply:
maybe add a todo